cluster-autoscaler/clusterstate/clusterstate.go

```diff
@@ -0,0 +1,337 @@
+		config:                config,
+		perNodeGroupReadiness: make(map[string]Readiness),
+		acceptableRanges:      make(map[string]AcceptableRange),
+	}
+}
+
+// RegisterScaleUp registers scale up.
+func (csr *ClusterStateRegistry) RegisterScaleUp(request *ScaleUpRequest) {
+	csr.Lock()
+	defer csr.Unlock()
+	csr.scaleUpRequests = append(csr.scaleUpRequests, request)
+}
+
+// RegisterNodeDeletion registers node deletion.
+func (csr *ClusterStateRegistry) RegisterNodeDeletion(request *NodeDeletionRequest) {
+	csr.Lock()
+	defer csr.Unlock()
+	csr.nodeDeletionRequests = append(csr.nodeDeletionRequests, request)
+}
+
+// To be executed under a lock.
+func (csr *ClusterStateRegistry) cleanUp(currentTime time.Time) {
+	newSur := make([]*ScaleUpRequest, 0)
+	for _, sur := range csr.scaleUpRequests {
+		if sur.ExpectedAddTime.After(currentTime) {
+			newSur = append(newSur, sur)
+		}
+	}
+	csr.scaleUpRequests = newSur
+
+	newNdr := make([]*NodeDeletionRequest, 0)
+	for _, ndr := range csr.nodeDeletionRequests {
+		if ndr.ExpectedDeleteTime.After(currentTime) {
+			newNdr = append(newNdr, ndr)
+		}
+	}
+}
+
+// UpdateNodes updates the state of the nodes in the ClusterStateRegistry and recalculates the statss
+func (csr *ClusterStateRegistry) UpdateNodes(nodes []*apiv1.Node, currentTime time.Time) error {
+	targetSizes, err := getTargetSizes(csr.cloudProvider)
+	if err != nil {
+		return err
+	}
+
+	csr.Lock()
+	defer csr.Unlock()
+
+	csr.cleanUp(currentTime)
+	csr.nodes = nodes
+	csr.perNodeGroupReadiness, csr.totalReadiness = csr.calculateReadinessStats(currentTime)
+	csr.acceptableRanges = csr.calculateAcceptableRanges(targetSizes)
+	return nil
+}
+
+// getTargetSizes gets target sizes of node groups.
+func getTargetSizes(cp cloudprovider.CloudProvider) (map[string]int, error) {
+	result := make(map[string]int)
+	for _, ng := range cp.NodeGroups() {
+		size, err := ng.TargetSize()
+		if err != nil {
+			return map[string]int{}, err
+		}
+		result[ng.Id()] = size
+	}
+	return result, nil
+}
+
+// IsClusterHealthy returns true if the cluster health is within the acceptable limits
+func (csr *ClusterStateRegistry) IsClusterHealthy(currentTime time.Time) bool {
+	csr.Lock()
+	defer csr.Unlock()
+
+	totalUnready := csr.totalReadiness.Unready + csr.totalReadiness.LongNotStarted
+
+	if totalUnready > csr.config.OkTotalUnreadyCount &&
+		float64(totalUnready) > csr.config.MaxTotalUnreadyPercentage/100.0*float64(len(csr.nodes)) {
+		return false
+	}
+
+	return true
+}
+
+// IsNodeGroupHealthy returns true if the node group health is within the acceptable limits
+func (csr *ClusterStateRegistry) IsNodeGroupHealthy(nodeGroupName string) bool {
+	readiness, found := csr.perNodeGroupReadiness[nodeGroupName]
+	if !found {
+		glog.Warningf("Failed to find readiness information for %v", nodeGroupName)
+		return false
+	}
+	acceptable, found := csr.acceptableRanges[nodeGroupName]
+	if !found {
+		glog.Warningf("Failed to find acceptable ranges for %v", nodeGroupName)
```

Comment:
Shouldn't we return here? it seems that if acceptable is empty we carry on with the execution.